package.json

```diff
@@ -4,7 +4,7 @@
   "private": true,
   "dependencies": {
     "@nervosnetwork/ckb-sdk-utils": "0.103.1",
-    "antd": "4.16.10",
+    "antd": "^4.16.10",
```

Comment:
Dependencies should be pinned and updated explicitly so we can know the changes introduced by each upgrade.

Reply:
> Dependencies should be pinned and updated explicitly so we can know the changes introduced by each upgrade.

If you expect accurate dependency version changes, you should do so by comparing yarn.lock (although it is less readable, there are tools we can use, such as yarn-lock-diff), Only compare the differences in package.json to obtain a part of the change, and cannot obtain deeper descendants dependencies.

I prefer to write a variable scope in package.json, beacause yarn.lock exists, this does not break consistency across multiple runs, and the main thing is to reduce duplicate versions.
for example:
Suppose we pinned dependencies `a@1.0.0`, `b@1.0.0`(dep `a@^1.0.1`), `c@1.0.0`(dep `a@^1.0.2`).
`a@1.0.0` will be hoist to the top layer of node_modules tree because of the dependencies of the root level.
Although the version that `b` and `c` depends on each version can share `a@1.0.2`, the `a@1.0.0` we dependent on it cannot be given the top level, so `b` and `c` have to create a new node_modules in their own directory, now we have three `a` package.
If you change `a@1.0.0` to `a@^1.0.0`, there will only be one `a@1.0.2` at the top of the node modules tree.

Different packages of different versions sometimes bring a lot of trouble, so it is best to avoid this.

Reply:
> > Dependencies should be pinned and updated explicitly so we can know the changes introduced by each upgrade.
> 
> If you expect accurate dependency version changes, you should do so by comparing yarn.lock (although it is less readable, there are tools we can use, such as yarn-lock-diff), Only compare the differences in package.json to obtain a part of the change, and cannot obtain deeper descendants dependencies.
> 
> I prefer to write a semver in package.json, beacause yarn.lock exists, this does not break consistency across multiple runs, and the main thing is to reduce duplicate versions. for example: Suppose we pinned dependencies `a@1.0.0`, `b@1.0.0`(dep `a@^1.0.1`), `c@1.0.0`(dep `a@^1.0.2`). `a@1.0.0` will be hoist to the top layer of node_modules tree because of the dependencies of the root level. Although the version that `b` and `c` depends on each version can share `a@1.0.2`, the `a@1.0.0` we dependent on it cannot be given the top level, so `b` and `c` have to create a new node_modules in their own directory, now we have three `a` package. If you change `a@1.0.0` to `a@^1.0.0`, there will only be one `a@1.0.2` at the top of the node modules tree.
> 
> Different packages of different versions sometimes bring a lot of trouble, so it is best to avoid this.

Relying on `lock` while keeping versions unpinned imposes a burden on developers to debug when some upgrades are breaking, as mentioned in https://docs.renovatebot.com/dependency-pinning/#what-a-lock-file-doesnt-do-for-you

In the case mentioned above, I would prefer to keep `a@1.0.0` apart from `a@1.0.2`. It just uses more space but leaves the dependencies explicit.

Reply:
> Relying on lock while keeping versions unpinned imposes a burden on developers to debug when some upgrades are breaking, as mentioned in https://docs.renovatebot.com/dependency-pinning/#what-a-lock-file-doesnt-do-for-you

I think that when an upgrade dependency package is error, there is no significant difference in debugging complexity between projects that are pinned versions and semver verions used.

I read the article in the link, this article is aimed primarily at users of tools like Renovate, and many of the arguments it makes in the first half of the article are based on the absence of lock files (similar to the published package).
Some of them I think are worth talking about:
> The lock file has only delayed the inevitable problem, and provides much less visibility than package.json, because it's not designed to be human readable and is quite dense.
1. Lock files are less readable, this is correct, but we generally use tools to access it, such as `yarn why`.
> Alternatively, maybe someone thinks "This lockfile is probably really out of date and might be missing some essential patches" and decides to yarn upgrade the whole thing in one go. No doubt the diff will be full of green and red as many direct and indirect dependencies will have changed versions. Then it's pushed to CI for testing, fails, and you have to guess which of the changes caused it until you eventually narrow it down to foobar. This might require even manually looking through the lock file diffs line by line. Maybe dep blahblah also broke at the same time, to make it even harder.
2. When trying to upgrade all dependencies at once, whether pinned versions or semver versions, the complexity should be the same.
> By ceding control of direct dependency versions to the lock file, you have lost the ability to know when things are updated. You also may be missing out on really important patches you're not even aware of, because they're "in range" yet locked back to vulnerable or buggy versions in the lock file.
3. This is actually to show that the use of pinned versions in package.json will have more accurate root-level dependencies display, but semver versions can also achieve consistent results without much cost.

There is a yarn behavior to explain:
When root is dependent on a low version of `a@^1.0.0(locked 1.0.0)`, adding a second `b@^1.0.0(dep a@^1.0.1)` does not result in `a@^1.0.0` for root The locked version is automatically upgraded, but two packages will appear, so don't worry about the automatic upgrade behavior with semver.

Reply:
Compared to the benefits brought by pinned versions (I haven't thought of any benefits in this project), It leads to more package version conflicts, which can do a lot of harm.

The most common is the type inconsistency error caused by inconsistent package version, which requires debugging for a period of time to find the version problem, and the other is the direct data transfer between different versions of the package call interface caused by some unexpected bugs, some are hidden, difficult to detect.

For example, during this upgrade PR, I encountered the inconsistency of the version of `@babel/core`, a high version of `@babel/core` parsed by a (AST?) node, the `@babel/traverse` introduced by the lower version of `@babel/core` does not recognize, and then build an error.

Reply:
The conflict may be introduced by pinned dependencies and also has a counterpart of semver dependencies.

Say the case of `a@^1.0.0`, `b@1.0.0(dep a@^1.0.1)` and `c@1.0.0(dep a@^1.0.2)`. If `a@1.0.2` breaks `b@1.0.0`, it would be cumbersome to figure out.

But errors introduced by dependencies are easy to fix, just takes time. For me, the most important point is clarity. We should know what codes we are going to rely on and upgrade them confidently with a peek at their changes.

Reply:
> errors introduced by dependencies are easy to fix

That may be because we have different ideas about how much it costs to fix a bug when multiple versions of the same package conflict.

It may take more time to continue discussing this cost. I can change it back to pinned versions first, and then experience it for a while.